### DISCUSSION TITLE
Removing a defect from a batch of images

### DISCUSSION THREAD

**Discussion by MarcoLeoncino** [ImageMagick/ImageMagick#5442] [Digital Image Processing] [answered]
Dear all, 

I am totally new to ImageMagick and I would like to ask you your support/suggestions.
Is it possible to remove a defect from a batch of _n_ images? A typical example is shown in the attached picture, highlighted in the red rectangle:

![Test_001](https://user-images.githubusercontent.com/111227897/184534573-02fe355e-cd68-48e7-a945-6ff66ecdadac.jpg)

it is an hair which was present in the 16mm camera gate when the film was taken. I built a film scanner which takes a picture of each frame. All frames are then merged with ffmpeg to obtain an avi output. The hair is more or less always stable in the same position.
And no, unfortunately I cannot remove the defect with a video processing software like Davinci. 

If I could remove such defects, frame by frame, in a more or less automatic view, before combining them... I would be extremely happy. I was thinking to create a mask, which contains the hair only, then subtract the mask from each image. But I am too unexperienced to this.

Thanks a lot for your time!

**Comment by snibgo** [comment]
What version of IM, on what platform? I will assume IM v7 on Windows 8.1.

There are many methods for this. I suggest:

1. Find a crop that includes the entire hair, plus a bit more. (I include a bit less, to avoid your red line.) 

2. Find where the hair is. I find where the difference between the image and a 7x7 median is more that 5%. This makes a mask that also includes other non-hair pixels, but that's okay.

3. Make the masked pixels transparent.

4. Fill in transparent pixels (aka "inpainting"). The method I show here fills pixels by copying adjacent opaque pixels.

5. Composite the result back in the original frame.

For doing multiple frames, put this in a loop.

Windows BAT script:
```
magick ^
  hairInGate.jpg ^
  -crop 103x76+593+10 +repage ^
  +write mpr:ONEHAIR ^
  +write x.png ^
  ( +clone ^
    -statistic median 7x7 ^
    +write x1.png ^
  ) ^
  -compose Difference -composite ^
  +write d.png ^
  -threshold 5%% ^
  -alpha off ^
  -morphology dilate disk:1 ^
  -negate ^
  +write m.png ^
  mpr:ONEHAIR ^
  +swap ^
  -compose CopyOpacity -composite ^
  t.png

call %PICTBAT%shiftFill t.png 10x10

magick ^
  hairInGate.jpg ^
  t_sf.png ^
  -geometry +593+10 ^
  -compose Over -composite ^
  out.png
```
This takes about 6 seconds. Each `+write x.png` etc is purely for debugging, so we can see intermediate results. They can be removed for better performance.

![out](https://user-images.githubusercontent.com/33812368/184541619-d6d892c1-fa45-4a6e-b308-3ac2cd7cc7ec.png)

**Comment by MarcoLeoncino** [comment]
Hi,   

thank you for your fast reply! What can I say, I am very surprised about your code, for me it would have been impossible to have created something like this! 

Actually, I am running WIN11, I created a text file, in which I copied your code, replaced all the "hairInGate.jpg" with the file name which I have, renamed the file into . bat and run it.


I got the following output files:

![x1](https://user-images.githubusercontent.com/111227897/184543628-49595a4d-c761-4fd9-9ea0-178473fcd2e1.png)
![d](https://user-images.githubusercontent.com/111227897/184543622-fbab1b62-f128-4a42-ac20-845ca3ac7fcc.png)
![m](https://user-images.githubusercontent.com/111227897/184543625-25e8e14b-a28b-4532-a85b-6e4569125c2c.png)
![t](https://user-images.githubusercontent.com/111227897/184543626-caa84c61-19e3-4ead-bfca-b3392b2b0a48.png)
![x](https://user-images.githubusercontent.com/111227897/184543627-76f8fc1f-8223-4f85-bf16-925a86940a36.png)

but I cannot find the out.png which shall be the corrected file.
Running in the powershell, I think that the problem is  the "t_sf.png ^" file which cannot be found. 
Was it a typo?

How can I make a loop to correct many files?

Thanks a lot!

**Comment by snibgo** [comment]
I know nothing about Powershell.

I expect your script also failed at...
```
call %PICTBAT%shiftFill t.png 10x10
```

The script `shiftFill.bat` creates and runs another script. `shiftFill.bat` is available at [Filling holes](http://im.snibgo.com/fillholes.htm). However, I expect to update that page in the next 12 hours.

For now, you can replace...
```
call %PICTBAT%shiftFill t.png 10x10
```
... with ...
```
magick -script sf_tmp.scr
```
... which needs a text file called sf_tmp.scr that contains this:
```
t.png +write mpr:ORIG -compose DstOver 
mpr:ORIG -geometry +1+0 -composite 
mpr:ORIG -geometry -1+0 -composite 
mpr:ORIG -geometry +0+1 -composite 
mpr:ORIG -geometry +0-1 -composite 
mpr:ORIG -geometry +2+0 -composite 
mpr:ORIG -geometry -2+0 -composite 
mpr:ORIG -geometry +0+2 -composite 
mpr:ORIG -geometry +0-2 -composite 
mpr:ORIG -geometry +3+0 -composite 
mpr:ORIG -geometry -3+0 -composite 
mpr:ORIG -geometry +0+3 -composite 
mpr:ORIG -geometry +0-3 -composite 
mpr:ORIG -geometry +4+0 -composite 
mpr:ORIG -geometry -4+0 -composite 
mpr:ORIG -geometry +0+4 -composite 
mpr:ORIG -geometry +0-4 -composite 
mpr:ORIG -geometry +5+0 -composite 
mpr:ORIG -geometry -5+0 -composite 
mpr:ORIG -geometry +0+5 -composite 
mpr:ORIG -geometry +0-5 -composite 
mpr:ORIG -geometry +6+0 -composite 
mpr:ORIG -geometry -6+0 -composite 
mpr:ORIG -geometry +0+6 -composite 
mpr:ORIG -geometry +0-6 -composite 
mpr:ORIG -geometry +7+0 -composite 
mpr:ORIG -geometry -7+0 -composite 
mpr:ORIG -geometry +0+7 -composite 
mpr:ORIG -geometry +0-7 -composite 
mpr:ORIG -geometry +8+0 -composite 
mpr:ORIG -geometry -8+0 -composite 
mpr:ORIG -geometry +0+8 -composite 
mpr:ORIG -geometry +0-8 -composite 
mpr:ORIG -geometry +9+0 -composite 
mpr:ORIG -geometry -9+0 -composite 
mpr:ORIG -geometry +0+9 -composite 
mpr:ORIG -geometry +0-9 -composite 
mpr:ORIG -geometry +10+0 -composite 
mpr:ORIG -geometry -10+0 -composite 
mpr:ORIG -geometry +0+10 -composite 
mpr:ORIG -geometry +0-10 -composite 
 mpr:ORIG -background White -compose Over -composite 
+write t_sf.png 
-exit 
```

> How can I make a loop to correct many files?

I suggest a Windows "for" loop. At the command prompt, type "help for".

**Comment by MarcoLeoncino** [comment]
Thanks a lot, now it worked well!
This is my output:

![out](https://user-images.githubusercontent.com/111227897/184547703-9f1c49fc-7a83-4063-bd8b-3f58044f0dc2.png)
 
Can you please tell me more about shiftFill.bat? After it will be in the repository, in 12 hours, what shall I do? Where will I find that bat file and where shall I save it?

Thanks again a lot!

**Comment by snibgo** [comment]
> Can you please tell me more about shiftFill.bat? After it will be in the repository, in 12 hours, what shall I do? Where will I find that bat file and where shall I save it?

As I said, `shiftFill.bat` is available at [Filling holes](http://im.snibgo.com/fillholes.htm). It is also documented on that page, with illustrated examples. You can copy/paste the script from that page.

See also [Zipped BAT files](http://im.snibgo.com/zipbats.htm) which describes my script files generally, with instructions on using them, and a zip file of all my published scripts.

**Comment by MarcoLeoncino** [comment]
You are totally right, I am sorry. I shall read the description more carefully.
However, your work is simply amazing.
Thanks a lot again!

**Comment by MarcoLeoncino** [comment]
It´s very strange. I was adding the shiftFill.bat and it was working well. Now it seems that I get again the previous error: 

magick: unable to open image 't_sf.png': No such file or directory @ error/blob.c/OpenBlob/3527

**Comment by MarcoLeoncino** [comment]
I summarize what I am doing:

in the same folder I have the "Test.bat" file which contains:

magick ^
  Imre_reel2_001.jpg ^
  -crop 103x76+593+10 +repage ^
  +write mpr:ONEHAIR ^
  +write x.png ^
  ( +clone ^
    -statistic median 7x7 ^
    +write x1.png ^
  ) ^
  -compose Difference -composite ^
  +write d.png ^
  -threshold 5%% ^
  -alpha off ^
  -morphology dilate disk:1 ^
  -negate ^
  +write m.png ^
  mpr:ONEHAIR ^
  +swap ^
  -compose CopyOpacity -composite ^
  t.png

call %PICTBAT%shiftFill t.png 10x10

magick ^
  Imre_reel2_001.jpg ^
  t_sf.png ^
  -geometry +593+10 ^
  -compose Over -composite ^
  out.png


In the same folder I have the picture to be modified and the shiftFill.bat file.
However it seems that there are still problems with the 't_sf.png' file...

**Comment by snibgo** [comment]
Have you looked at shiftFill.bat? It calls other scripts. Have you downloaded them?

**Comment by MarcoLeoncino** [comment]
Correct.

I think I have to add:

call %PICTBAT%**setInOut** %1 sf
call **parseXxY2** %WW% %HH% sfXY %MAX_SHFT%
call %PICTBAT%**getLtrb** %sfLTRB%
if "%IM_VSN_6or7%"=="" call %PICTBAT%**isIm6or7**

Am I right?

**Reply by MarcoLeoncino** [reply]
Hi, 

I have downloaded all the above scripts in bold, but  when I try to search for isIm6or7 from here:

http://im.snibgo.com/scrwscr.htm

I get a 404 Not Found

Can you please check?

I am also struggling on how to run the for loop on many and many files contained in a directory and to save them in an another directory.
I am sorry, I am not a good programmer!


Edit: I have also installed echoOffSave.bat and echoRestore.bat

**Comment by MarcoLeoncino** [comment]
And this is the content of my directory: 


![Immagine 2022-08-15 183618](https://user-images.githubusercontent.com/111227897/184676868-6649a6f2-eabe-4a4b-b5ee-0346c2166d8e.jpg)

**Comment by snibgo** [comment]
isIm6or7.bat is:
```
for /F "usebackq tokens=2" %%L in (`%IMG7%magick identify -format "%%[version]" xc: `) do set IM_VSN=%%L

set IM_VSN_6or7=%IM_VSN:~0,1%

echo IM_VSN=%IM_VSN% IM_VSN_6or7=%IM_VSN_6or7%
```
If you only run v7, you could shorten the script to:
```
set IM_VSN_6or7=7
```

Some of my scripts only work for v6, some only work for v7, and some take different actions depending on whether IM is v6 or v7.

I am gradually changing this so they will all work for v7.

On loops: once you have a script that works for one file (one frame), you can call it from another script that loops through files (all frames). Something like this:
```
for %%F in (Imre_reel*.jpg) do call myscript %%F
```

**Reply by MarcoLeoncino** [reply]
Thanks a lot again for your help, however I still have some issues. This is the output of my terminal.
Dehair.bat is the script which calls all the others.

D:\Digi\Preliminary_Tests\Test_005> .\Dehair.bat

D:\Digi\Preliminary_Tests\Test_005>magick   Imre_reel2_001.JPG   -crop 103x76+593+10 +repage   +write mpr:ONEHAIR   +write x.png   ( +clone     -statistic median 7x7     +write x1.png   )   -compose Difference -composite   +write d.png   -threshold 5%   -alpha off   -morphology dilate disk:1   -negate   +write m.png   mpr:ONEHAIR   +swap   -compose CopyOpacity -composite   t.png

D:\Digi\Preliminary_Tests\Test_005>call shiftFill t.png 10x10
"!" non è riconosciuto come comando interno o esterno,
 un programma eseguibile o un file batch.
shiftFill: MAX_SHFT_W=10 MAX_SHFT_H=10
shiftFill: nITER=11
IM_VSN=7.1.0-45 IM_VSN_6or7=7
ECHO disattivato.
magick: unable to open image 't_sf.png': No such file or directory @ error/blob.c/OpenBlob/3568.

_"!" non è riconosciuto come comando interno o esterno, un programma eseguibile o un file batch._

**means**: "!" is not recognized as an internal or external command, an executable program or a batch file.

**Comment by snibgo** [comment]
In shiftfill.bat, change the line near the top:
```
@setlocal
```
... to:
```
@setlocal enabledelayedexpansion
```

**Comment by MarcoLeoncino** [comment]
Wow... thanks a lot for the super fast reply. Now I am able to find the output of the program, it is fantastic. I am again impressed about your capabilities to program.

**Comment by MarcoLeoncino** [comment]
Thanks for the for loop too, the issue I am finding is that the input and output filenames are hardcoded in the main script.
Therefore I must find a way not to harcode them, in order to let the main script running on all files... and since I don´t know batch programming, this is not trivial for me.

**Comment by snibgo** [comment]
Your script Dehair.bat includes hardcoded filenames such as  Imre_reel2_001.JPG for the input. Instead it can use numbered parameters such as %1 for the first parameter. Then you can...
```
call Dehair.bat Imre_reel2_001.JPG
```
But that is a question about scripting, not ImageMagick.

**Reply by MarcoLeoncino** [reply]
Thank you, but as you could have understood I am a null in programming... 
If I understand, in my script I shall replace, the two times when it appears, the following line: 

Imre_reel2_001.JPG --> Imre_reel2_%1.JPG

Then I can make another *bat file which calls the script as many times as the number of file which I have.
But I cannot figure out how...

**Comment by MarcoLeoncino** [comment]
Can you please give me an help on this? I would be really grateful for that...

**Comment by snibgo** [comment, marked as answer]
As I have said, you could have a script, which might be called `deHairAll.bat`, like this:
```
for %%F in (Imre_reel*.jpg) do call deHair.bat %%F
```
So for each file that matches "Imre_reel*.jpg", it calls deHair.bat, giving it the filename.

The script file `deHair.bat` would use "%1" instead of the input filename, so it could be:
```
%IMG7%magick ^
  %1 ^
  -crop 103x76+593+10 +repage ^
  +write mpr:ONEHAIR ^
  ( +clone ^
    -statistic median 7x7 ^
  ) ^
  -compose Difference -composite ^
  -threshold 5%% ^
  -alpha off ^
  -morphology dilate disk:1 ^
  -negate ^
  mpr:ONEHAIR ^
  +swap ^
  -compose CopyOpacity -composite ^
  t.png

call %PICTBAT%blurFill t.png

call %PICTBAT%shiftFill t.png 10x10

%IMG7%magick ^
  %1 ^
  t_sf.png ^
  -geometry +593+10 ^
  -compose Over -composite ^
  outdir\%~n1.jpg
```
In each `magick` command, the input file is `%1`. The overall output is `outdir\%~n1.jpg`, so it writes to a different directory (that must already exist).

**Comment by MarcoLeoncino** [comment]
Thanks a lot! It works!

**Comment by MarcoLeoncino** [comment]
I have a furher question, sorry! I have noticed that you have added the blurFIll procedure in the list. Therefore I have saved it in my folder and I also added the .bat which this scripts calls (quantFp.bat). In the blurFIll.bat I have also replaced 

@setlocal

... to:

@setlocal enabledelayedexpansion

but I still have this error (or warning?) line:

magick: unable to open image '\temp\bf_tmpin.miff': No such file or directory @ error/blob.c/OpenBlob/3568. 

Is that normal? Even if I get that message, the program goes up to the end and saves the pictures.

Now I am also looking to your other scripts and hotPixels.bat could be interesting to remove some spots. But I have to check if this method is effective also with black and white only pictures...

**Comment by snibgo** [comment]
blurFill.bat doesn't need enabledelayedexpansion, but I expect adding that is harmless.

> magick: unable to open image '\temp\bf_tmpin.miff': No such file or directory

Do you have a directory named "\temp\"? If you don't, then that is the problem. You can add that directory with:
```
md \temp
```

**Reply by MarcoLeoncino** [reply]
Thanks, but where shall the directory \temp be created?

**Comment by snibgo** [comment]
I don't understand the question.

**Reply by MarcoLeoncino** [reply]
You have asked if I have the \temp directory. But where shall this directory be made? In the directory where I have installed ImageMagick? Or in the folder where I have the scripts?

**Reply by MarcoLeoncino** [reply]
Or perhaps you mean that the line md \temp (make directory) shall be inserted in a script?

**Comment by snibgo** [comment]
The command ...
```
md \temp
```
... will make a directory (aka "folder") called `temp` in the root directory of the current drive. The backslash at the front puts it in the root, not in any subdirectory. So a `magick` command like ...
```
magick toes.png \temp\out.miff
```
... will create a file in that directory.

**Comment by MarcoLeoncino** [comment]
Hi Snibgo,

I was working a bit to improve the scanning technique and to start a standardized procedure.
In particular to tune up the macro you wrote for me time ago to adapt as best as possible on all the frame.
I am pretty happy, but I was wondering if you could give me an advice to reduce the transition to the area of the picture which is not processed from the one on which the processed image is overlapped.

Here you are an enlargement of the region where you can find the transition:

![Immagine 2022-10-16 185257](https://user-images.githubusercontent.com/111227897/196048074-1a7a71a8-0cc1-4a38-a6db-1e4876deea7a.jpg)
 
At the moment the following are the settings used:

%IMG7%magick ^
  %1 ^
  -crop 290x280+1925+0 +repage ^
  +write mpr:ONEHAIR ^
  ( +clone ^
    -statistic median 35x35 ^
  ) ^
  -compose Difference -composite ^
  -threshold 5%% ^
  -alpha off ^
  -morphology dilate disk:1.05 ^
  -negate ^
  mpr:ONEHAIR ^
  +swap ^
  -compose CopyOpacity -composite ^
  t.png

call %PICTBAT%blurFill t.png
call %PICTBAT%shiftFill t.png 15x15

%IMG7%magick ^
  %1 ^
  t_sf.png ^
  -geometry +1925+0 ^
  -compose Over -composite ^
  Processed_pass1\%~n1.jpg


This is the raw image:

![Imre_reel2_00658](https://user-images.githubusercontent.com/111227897/196048191-ba1ada53-2397-48a1-8974-07214e393bfd.JPG)

And this is the processed image, using the settings above:

![Imre_reel2_00658](https://user-images.githubusercontent.com/111227897/196048206-a0d6092c-c4bc-4ca6-ba9a-4e3028346d2b.jpg)

I would like to know if I can do better...
Thanks a lot!

**Comment by MarcoLeoncino** [comment]
In other words, I was wondering if it would be possible to make the processed rectangle "less visible". Perhaps by adding an unfocused mask to that area?

**Comment by MarcoLeoncino** [comment]
Something like the following. 

![001](https://user-images.githubusercontent.com/111227897/196214250-49cf5340-affa-47f4-9bd9-6fda828b7454.jpg)

Here I have added the line:   -unsharp 2x2+2+0 ^
after   -compose CopyOpacity -composite ^

![002](https://user-images.githubusercontent.com/111227897/196214320-7ee68de4-79c6-46b2-b613-4f2a444d21d7.jpg)

what do you think? I am just trying to recreating the film grain in the processed area...

**Comment by MarcoLeoncino** [comment]
Instead, here, I put the unsharp command before the line:   -compose CopyOpacity -composite ^

![003](https://user-images.githubusercontent.com/111227897/196219843-4ead749c-947a-4138-9d13-9419929fd3f0.jpg)

**Comment by fmw42** [comment]
Sorry, but I do not understand your question.  Where is the region you are interested in?  What do you want do do there?

**Reply by MarcoLeoncino** [reply]
The discussion is long.

With an home made telecine machine (https://www.youtube.com/watch?v=Q2e5ctuxbUw), I scan a 16 mm film, frame by frame. Then I merge all the frames to recreate a movie. However, there are some complications.

In the rectangle you can see the defect (hair in the gate of a film camera) which Snibgo removes with his macro:

![196048191-ba1ada53-2397-48a1-8974-07214e393bfd](https://user-images.githubusercontent.com/111227897/196223987-13809efa-a5dc-4255-8769-127059d2fb76.jpg)

This is the result after the processing:

![196048206-a0d6092c-c4bc-4ca6-ba9a-4e3028346d2b](https://user-images.githubusercontent.com/111227897/196224052-410a0440-72c7-4338-a3b5-61db70aff153.jpg)

The macro is extremely powerful and can be run on hundreds of frames. I am really grateful to Snibgo for that.
But if you enlarge the repaired area you will see the following:

![196048074-1a7a71a8-0cc1-4a38-a6db-1e4876deea7a](https://user-images.githubusercontent.com/111227897/196224240-22b6b627-a256-4dd7-9f8a-bee7a6415767.jpg)

Basically, you can see the area where the correction has been done, because it is blurred, compared to the rest of the frame.
When you put all the frames together, to make a film, with ffmpeg, you will notice the rectangle.
By applying an unsharped mask to the repaired area, I obtain the following:

![ff](https://user-images.githubusercontent.com/111227897/196224499-8ce2ac87-2e41-484e-90f0-a50b389588e6.jpg)

Which makes the whole process less visible.
I hope that Snibgo will see this latest posts, because I would like to know his opinion (does the unsharp mask makes sense or is it useless and are there other, better ways?).

**Comment by snibgo** [comment]
Sorry, I don't have enough time to devote to this problem. And I can't evaluate the results, which would be the movie before and after applying de-hairing.

The first image you showed had grain, but almost no image detail within the rectangle. The second image does have image detail.

The blurry rectangle occurs because many pixels that are actually "grain" were falsely identified as "hair". If a sharpening seems to fix the problem, then that is good.

A more accurate detector can be built that distinguished "hair" from "grain" (which has a higher frequency) and from "image detail" (which has lower frequency). Then the mask will detect just the "hair" pixels, and only those will be fixed.

**Comment by MarcoLeoncino** [comment]
Thank you for your reply. It would be really interesting to have that more accurate detector...
But I respect the fact that now you are busy. Thank you anyway...
